### PR TITLE
IndexedStateT

### PR DESCRIPTION
**Pull request by nomisRev**

### CODE REVIEW

File: modules/core/arrow-test/src/main/kotlin/arrow/test/laws/ApplicativeErrorLaws.kt [hunk 1]
```diff
@@ -23,7 +23,7 @@ object ApplicativeErrorLaws {
       Law("Applicative Error Laws: handle with for success") { AE.applicativeErrorHandleWithPure(EQERR) },
       Law("Applicative Error Laws: attempt for error") { AE.applicativeErrorAttemptError(EQ_EITHER) },
       Law("Applicative Error Laws: attempt for success") { AE.applicativeErrorAttemptSuccess(EQ_EITHER) },
-      Law("Applicative Error Laws: attempt fromEither consistent with pure") { AE.applicativeErrorAttemptFromEitherConsistentWithPure(EQ_EITHER) },
+      Law("Applicative Error Laws: attempt fromEither consistent with just") { AE.applicativeErrorAttemptFromEitherConsistentWithPure(EQ_EITHER) },
```
Comment (pakoito) [inline modules/core/arrow-test/src/main/kotlin/arrow/test/laws/ApplicativeErrorLaws.kt]: _wink_

File: modules/docs/arrow-docs/docs/docs/datatypes/state/README.md [hunk 2]
```diff
@@ -70,8 +67,6 @@ Our `pop` function takes a `Stack` and returns an updated `Stack` and a `String`
 Let's write a new version of `pop` and `push` using `State`:
 
 ```kotlin:ank:silent
-import arrow.*
```
Comment (pakoito) [inline modules/docs/arrow-docs/docs/docs/datatypes/state/README.md]: I can hear the perf improving from here.

File: modules/core/arrow-instances-data/src/main/kotlin/arrow/instances/statet.kt [hunk 3]
```diff
@@ -3,136 +3,143 @@ package arrow.instances
...
 @extension
-interface StateTFunctorInstance<F, S> : Functor<StateTPartialOf<F, S>> {
+interface IndexedStateTFunctorInstance<F, S> : Functor<IndexedStateTPartialOf<F, S, S>> {
 
   fun FF(): Functor<F>
 
-  override fun <A, B> Kind<StateTPartialOf<F, S>, A>.map(f: (A) -> B): StateT<F, S, B> =
-    fix().map(FF(), f)
+  override fun <A, B> Kind<IndexedStateTPartialOf<F, S, S>, A>.map(f: (A) -> B): Kind<IndexedStateTPartialOf<F, S, S>, B> =
+    this.fix().map(FF(), f)
```
Comment (pakoito) [inline modules/core/arrow-instances-data/src/main/kotlin/arrow/instances/statet.kt]: This `this` shouldn't be necessary.

File: modules/docs/arrow-docs/docs/docs/datatypes/statet/README.md [hunk 4]
```diff
@@ -136,7 +136,7 @@ import arrow.instances.either.monadError.*
 
 fun stackOperationsS2() = 
  ForStateT<EitherPartialOf<StackError>, Stack, StackError>(Either.monadError<StackError>()) extensions {
-  binding {
+  binding<StateTPartialOf<EitherPartialOf<StackError>, Stack>, String> {
```
Comment (nomisRev) [inline modules/docs/arrow-docs/docs/docs/datatypes/statet/README.md]: @pakoito seems I this message didn't go through or I just forgot 🤔 

Anyways, the inference trips here. Same for `StateApi.monad<Stack>().binding { }`. Also IntelliJ doesn't complain and the user ends up with a `Kotlinc Back-end error..". Thoughts?

Comment (pakoito) [reply modules/docs/arrow-docs/docs/docs/datatypes/statet/README.md]: >Also IntelliJ doesn't complain and the user ends up with a `Kotlinc Back-end error..". Thoughts?

Async code and specially coroutines don't work on Ank. Hardcode the result instead.

Comment (nomisRev) [reply modules/docs/arrow-docs/docs/docs/datatypes/statet/README.md]: @pakoito

Comment (pakoito) [reply modules/docs/arrow-docs/docs/docs/datatypes/statet/README.md]: If this is blocking you, hardcode the result.

Comment (nomisRev) [reply modules/docs/arrow-docs/docs/docs/datatypes/statet/README.md]: @pakoito I think I am not explaining it properly 😅

In Arrow 0.7.3
* `binding { ... }` compiles and the generic types `<F, B>` are correctly inferred.
* `State<S, A>` is an alias for `StateT<ForId, S, A>`

In Arrow 0.8.0
* `binding { ... }` **does not compile**, the compiler trips over `<F, B>` with a vague `kotlinc Back-end error..` message. This is fixed by explicitly specifying the generic types but IDEA marks them as redundant. (gray & underlined).
* `State<S, A>` is an alias for `IndexedStateT<ForId, S, S, A>`
* `IndexedState<ForId, SA, SB, A>` is an alias for `IndexedStateT<ForId, SA, SB, A>`
* `StateT<F, S, A>` is an alias for `IndexedStateT<F, S, S, A>`

In other words, this results in a very tedious migration for `State` from `0.7.3` to `0.8.0` since it results in a `kotlinc Back-end error..` and IDEA claims the types do not need to be specified for `binding`.

**Comment by nomisRev** [comment]
@arrow-kt/maintainers I will close this PR while awaiting a fix. https://youtrack.jetbrains.com/issue/KT-28236
